File: read_write_by_binary_chunks.py
# -*- coding: utf-8 -*-
#!/usr/bin/python
import sys
import os
MAX_CHUNK_SIZE = 1024 # 1k
def copy (target_name, data, chunk_size, chunk_index, last_chunk):
    seek_start = chunk_size * chunk_index
    print len (data)
    #print seek_start
    with open(target_name, 'ab') as f :
        f.seek (0, seek_start)
        f.write(data)
        f.close ()

def read_and_copy (filename, target_name):
    chunk_index = 0
    last_chunk = False
    print filename
    # must use rb !
    with open(filename, "rb") as file :
        file.seek(0)
        while (not last_chunk):
            data = file.read (MAX_CHUNK_SIZE)
            #print chunk_index
            if len(data) == 0 :
                last_chunk = True
            copy (target_name, data, MAX_CHUNK_SIZE, chunk_index, last_chunk)
            chunk_index = chunk_index + 1
        file.close()
    print chunk_index



if __name__ == "__main__":
    length = len(sys.argv)
    if length < 2:
        raise Exception("Error no filename specified !")
    filename = sys.argv [1]
    target_name = filename + '_2'
    if os.path.isfile(target_name):
        response = raw_input("file %s already exists, would you like to override current file (y/yes): " % target_name)
        if (response == 'y' or response == 'yes'):
            os.remove(target_name)
            read_and_copy(filename, target_name)
    else :
        read_and_copy(filename, target_name)

File: read_write_by_binary_chunks.js
#!/usr/bin/env node
const meow = require('meow');
var path = require('path');
var fs = require('fs');

const CHUNCK_SIZE = 100;

const cli = meow(
    "push " +
    "$filename  " +
    "$host "
);
// console.log (cli.help);
var filename = cli.input[0];
var host = cli.input[1];
if (!filename || !host) {
    throw new Error('no host/filename specified !');
}
var fileDescriptor;
var totalChunks ;

/**
 *
 * @param buffer
 * @param chunkOffset
 * @param chunkSize
 * @param lastChunk
 */
var appendChunk = function (buffer, chunkOffset, chunkSize, lastChunk) {
    if (!fileDescriptor) {
        fileDescriptor = fs.openSync(filename + '_2', 'a');
        totalChunks = 0 ;
    }
    totalChunks++ ;
    // write to created file
    // fs.write(fd, buffer, offset, length[, position], callback)#
    fs.write(fileDescriptor, buffer, 0, buffer.length, (chunkOffset * buffer.length), function (err, num) {
        console.log(buffer.toString('utf-8', 0, num));
        /**
         *
         */
        if (lastChunk) {
            console.log('ended copied \n');
            console.log ('total chunks : ' + totalChunks);
            fs.close(fileDescriptor);
        }
    });

};
/**
 *
 */
fs.open(filename, 'r',
    /**
     *
     * @param status
     * @param fd
     */
    function (status, fd) {
        if (status) {
            console.log(status.message);
            return;
        }
        var buffer;
        var currentOffset = 0;
        /**
         * loop
         */
        (function continueReading() {
            buffer = new Buffer(CHUNCK_SIZE);
            fs.read(fd, buffer, 0, CHUNCK_SIZE, currentOffset * CHUNCK_SIZE, function (err, num) {
                //console.log(buffer.toString('utf-8', 0, num));
                //console.log(buffer);
                if (num < CHUNCK_SIZE) {
                    // we need only copy
                    appendChunk(buffer.slice(0, num), currentOffset++, CHUNCK_SIZE, true);
                }
                else {
                    appendChunk(buffer, currentOffset++, CHUNCK_SIZE);
                    continueReading();
                }
            });
        })();

    });
